pywbem_mock/_wbemconnection_mock.py

```diff
@@ -0,0 +1,1571 @@
+        iname.namespace = namespace
+
+        try:
+            self.instances[namespace]
+        except KeyError:
+            ce = CIMError(CIM_ERR_NOT_FOUND,
+                          'Namespace %s not found' % namespace)
+            raise ce
+
+        del_inst = None
+        for enum_tup in enumerate(self.instances[namespace]):
+            index = enum_tup[0]
+            inst = enum_tup[1]
+            if iname == inst.path:
+                if del_inst is not None:
+                    raise CIMError(CIM_ERR_FAILED, 'Invalid Repository. '
+                                   'Multiple instances with same path %s'
+                                   % inst.path)
+                # TODO confirm that we do this test on create and we
+                # can stop looking through whole list here.
+                else:
+                    del self.instances[namespace][index]
+                    del_inst = iname
+
+        if not del_inst:
+            raise CIMError(CIM_ERR_NOT_FOUND, 'Instance %s not found in '
+                                              'repository' % iname)
+
+    def _fake_enumerateinstances(self, namespace, **params):
+        """
+        Implements a server reactor for
+        :meth:`~pywbem.WBEMConnection.EnumerateInstances`.
+
+        Gets a list of subclasses if the classes exist in the repository
+        then executes getInstance for each to create the list of instances
+        to be returned.
+
+        Exceptions:
+            CIMError CIM_ERR_INVALID_NAMESPACE
+        """
+        # TODO we are duplicating the namespace test here.
+        self._validate_instances_namespace(namespace)
+
+        # TODO is there any case were we will get a string classname here???
+        cname_input = params['ClassName']
+        cname = cname_input.classname if isinstance(cname_input, CIMClassName) \
+            else cname_input
+
+        insts = self._enumerate_instances(cname, namespace, **params)
+
+        return self._make_tuple(insts)
+
+    def _fake_enumerateinstancenames(self, namespace, **params):
+        """
+            Implements a server reactor for
+            :meth:`~pywbem.WBEMConnection.EnumerateInstanceNames`
+
+            Get instance names for instances that match the path define
+            by `ClassName` and returns a list of the names.
+
+        """
+        self._validate_instances_namespace(namespace)
+
+        # TODO is there any case were we will get a string classname here???
+        cname_input = params['ClassName']
+        cname = cname_input.classname if isinstance(cname_input, CIMClassName) \
+            else cname_input
+
+        # TODO implement subclasses if class exists in namespace
+        # inst_paths = [inst.path for inst in self.instances[namespace]
+        #              if inst.path.classname == cname]
+
+        inst_paths = self._enumerate_instancenames(cname, namespace)
+        rtn_paths = [copy.copy(path) for path in inst_paths]
```

Comment:
We need to consolidate the copies that are being made, and to use that approach consistently across all operations. At the end, the caller of the fake operation should get a CIM object that is a copy of what is in the repository. Does `_make_tuple()` also make a copy?

Reply:
Mostly Done but marked TODO to verify.